File: apps/web/src/lib/search/chips.ts
/**
 * The chips: what Burro understood, one for each thing the spec holds.
 * docs/design/web.md, section 5.2.
 *
 * A chip is drawn from the spec the API returned, never from the edits. Its
 * words are names the API gave (a feature's label, a place's name) and site
 * copy for the codes. A part is marked "assumed" when nobody chose it.
 */

import { CRIME_ACCOUNT, crimeParts } from "@/content/crime";
import { CHIPS, PLACE } from "@/content/search";
import { MODE, SEGMENT, TENURE } from "@/content/labels";
import type {
  AreaSummary,
  AssumptionCode,
  MetaData,
  Operations,
  PreferenceSpec,
  Provenance,
  Tag,
  TagWeight,
} from "@/lib/api/schema";
import { grouped } from "@/lib/format";

import { counts } from "./counts";
import { edits } from "./edits";
import type { Assumed } from "./state";

export type ChipKind = "tenure" | "budget" | "place" | "feature" | "tag" | "area" | "usual";

export interface ChipPart {
  readonly text: string;
  /** True when nobody chose this part: the words did not say it. */
  readonly assumed: boolean;
  /**
   * True when the part is said in the row as well as in full, though it was assumed: a
   * word that has two meanings, and the one it was read as. It is never left for a press.
   */
  readonly always?: boolean;
}

export interface Chip {
  /** Tells one chip from another. It holds an id of the release, never anything typed. */
  readonly key: string;
  readonly kind: ChipKind;
  /** The place, feature, tag or area the chip is for. `null` where there is one of its kind. */
  readonly id: string | null;
  readonly label: string;
  readonly parts: readonly ChipPart[];
  /** True when the chip as a whole was assumed, or any part of it was. */
  readonly assumed: boolean;
  /** The edit that takes it out, where it can be taken out. */
  readonly removal: Operations | null;
  /** The edit that asks for the other end of a scale, where the chip is of one. */
  readonly turn: Operations | null;
}

const notChosen = (provenance: Provenance) => provenance === "default" || provenance === "inferred";

/**
 * The name of each place of the spec, as the answer that brought the spec gave it. Every
 * answer names every place of its spec. One that did not is said to have no name here, and
 * is never given a number or a name of the website's own.
 */
export function namesOfPlaces(
  spec: PreferenceSpec,
  placeNames: Readonly<Record<string, string>>,
): ReadonlyMap<string, string> {
  return new Map(
    spec.commutes.map((commute) => [commute.place_id, placeNames[commute.place_id] ?? PLACE.noName]),
  );
}

/** The name of the end of a scale that is asked for. `null` for a vibe that runs one way. */
export function endAskedFor(tag: Pick<Tag, "shape" | "low_end" | "high_end">, toward: TagWeight["toward"]) {
  if (tag.shape !== "scale") return null;
  return (toward === "low" ? tag.low_end : tag.high_end) ?? null;
}

/** A budget as the person gave it: pounds, and "a month" for a rent. */
export function budgetText(amount: number, tenure: PreferenceSpec["tenure"]): string {
  return tenure === "rent" ? `£${grouped(amount)} a month` : `£${grouped(amount)}`;
}

/** What the page knows of the search that the spec alone does not show. */
export interface Beside {
  /**
   * True when the person picked renting or buying before anything was sent. The page then
   * shows the default the API served for that tenure, which no edit made, so the spec
   * calls the tenure a default.
   */
  readonly tenurePicked?: boolean;
  /** The word each vibe was read from, where the word has two meanings: by the key of its chip. */
  readonly quoted?: Readonly<Record<string, string>>;
}

export function chipsOf(
  spec: PreferenceSpec,
  meta: Pick<MetaData, "features" | "tags">,
  areas: readonly AreaSummary[],
  placeNames: Readonly<Record<string, string>>,
  assumed: Assumed,
  beside: Beside = {},
): Chip[] {
  const has = (key: string, code: AssumptionCode) => assumed[key]?.includes(code) ?? false;
  const chip = (
    made: Omit<Chip, "assumed" | "parts" | "turn"> & { parts?: ChipPart[]; assumed?: boolean; turn?: Operations },
  ): Chip => {
    // A code the website has no word for makes a part with nothing in it, which is left out.
    const parts = (made.parts ?? []).filter((part) => Boolean(part.text));
    return {
      ...made,
      parts,
      assumed: (made.assumed ?? false) || parts.some((part) => part.assumed),
      turn: made.turn ?? null,
    };
  };
  const chips: Chip[] = [];

  // An entry of 0 is a thing a person took off. It is said to count for nothing, so that
  // "ignore the high street" is never shown as a wish for one, and it has nothing to remove.
  const takenOff: ChipPart[] = [{ text: CHIPS.off, assumed: false }];

  // What was asked for by way of character comes first: the vibes, and then the things
  // that count nearby. They are what sets one search apart from another, and were once
  // the part that was never seen, behind a tenure nobody said, a budget and a workplace.
  for (const weight of spec.tags) {
    const key = `tag:${weight.tag_id}`;
    const tag = meta.tags.find((one) => one.tag_id === weight.tag_id);
    const end = tag === undefined ? null : endAskedFor(tag, weight.toward);
    const name = tag?.label ?? weight.tag_id;
    // A word with two meanings is quoted, with what it was read as, until the person sets the vibe.
    const word = has(key, "word") ? beside.quoted?.[key] : undefined;
    const on = counts(weight);
    // Recorded crime counts only when a person asks for it. A vibe whose recipe holds it
    // says so on its chip, in the row, so that no search counts it unseen.
    const crime = tag !== undefined && crimeParts(tag, meta.features).length > 0;
    chips.push(
      chip({
        key,
        kind: "tag",
        id: weight.tag_id,
        // A scale says which end is asked for. A vibe that runs one way is asked for, or is not.
        label: end === null || !on ? name : CHIPS.towards(name, end),
        parts: on
          ? [
              ...(crime ? [{ text: CRIME_ACCOUNT.chip, assumed: false, always: true }] : []),
              ...(word === undefined ? [] : [{ text: CHIPS.readFrom(word), assumed: true, always: true }]),
            ]
          : takenOff,
        assumed: weight.provenance === "inferred" || has(key, "weight"),
        removal: on ? edits.tagOff(weight.tag_id) : null,
        ...(on && end !== null
          ? { turn: edits.tagWeight(weight.tag_id, weight.weight, weight.toward === "low" ? "high" : "low") }
          : {}),
      }),
    );
  }

  for (const weight of spec.weights) {
    if (weight.provenance === "default") continue;
    const metric = meta.features.find((feature) => feature.feature_id === weight.feature_id);
    const key = `feature:${weight.feature_id}`;
    const direction: ChipPart[] =
      metric?.polarity === "either"
        ? [
            {
              text: weight.direction === "more" ? CHIPS.more : CHIPS.fewer,
              assumed: has(key, "direction"),
            },
          ]
        : [];
    chips.push(
      chip({
        key,
        kind: "feature",
        id: weight.feature_id,
        // The plain name the API gives it, which says the wish where there is one way to wish.
        label: metric?.short_label ?? weight.feature_id,
        parts: counts(weight) ? direction : takenOff,
        assumed: weight.provenance === "inferred" || has(key, "weight"),
        removal: counts(weight) ? edits.featureOff(weight.feature_id) : null,
      }),
    );
  }

  // Then where the person must get to, and what they can pay.
  const names = namesOfPlaces(spec, placeNames);
  for (const commute of spec.commutes) {
    const key = `place:${commute.place_id}`;
    chips.push(
      chip({
        key,
        kind: "place",
        id: commute.place_id,
        label: names.get(commute.place_id) ?? "",
        parts: [
          { text: MODE[commute.mode], assumed: has(key, "mode") },
          { text: CHIPS.within(commute.max_minutes), assumed: has(key, "max_minutes") },
          {
            text: commute.strictness === "hard" ? CHIPS.firm : CHIPS.flexible,
            assumed: has(key, "strictness"),
          },
        ],
        removal: edits.placeRemove(commute.place_id),
      }),
    );
  }

  if (spec.budget.amount !== null) {
    chips.push(
      chip({
        key: "budget",
        kind: "budget",
        id: null,
        label: budgetText(spec.budget.amount, spec.tenure),
        parts: [
          { text: SEGMENT[spec.budget.segment], assumed: has("budget", "segment") },
          {
            text: spec.budget.strictness === "hard" ? CHIPS.firm : CHIPS.flexible,
            assumed: has("budget", "strictness"),
          },
        ],
        removal: edits.budgetClear(),
      }),
    );
  }

  for (const rule of spec.areas) {
    chips.push(
      chip({
        key: `area:${rule.area_id}`,
        kind: "area",
        id: rule.area_id,
        label: areas.find((area) => area.area_id === rule.area_id)?.name ?? rule.area_id,
        parts: [{ text: rule.rule === "exclude" ? CHIPS.hidden : CHIPS.only, assumed: false }],
        removal: edits.areaClear(rule.area_id),
      }),
    );
  }

  // Renting or buying is said of every search, and is most often what nobody said. It comes
  // after everything that was, and before the settings nobody chose, which come last.
  chips.push(
    chip({
      key: "tenure",
      kind: "tenure",
      id: null,
      label: TENURE[spec.tenure],
      assumed: beside.tenurePicked !== true && (notChosen(spec.tenure_from) || has("tenure", "tenure")),
      removal: null,
    }),
  );

  const usual = spec.weights.filter((weight) => weight.provenance === "default").length;
  if (usual > 0) {
    chips.push(
      chip({
        key: "usual",
        kind: "usual",
        id: null,
        label: CHIPS.usual(usual),
        assumed: true,
        removal: null,
      }),
    );
  }

  return chips;
}

File: apps/ios/BurroKit/Sources/BurroKit/Features/Search/Chips.swift
import Foundation

// The chips: what Burro understood, one for each thing the spec holds.
// docs/design/web.md, section 5.2. It mirrors apps/web/src/lib/search/chips.ts.
//
// A chip is drawn from the spec the API returned, never from the edits. Its
// words are names the API gave (a feature's label, a place's name) and the
// app's words for the codes. A part is marked "assumed" when nobody chose it.

/// One part of a chip: how a journey is made, or whether a limit is firm.
struct ChipPart: Hashable, Sendable {
    let text: String
    /// True when nobody chose this part: the words did not say it.
    let assumed: Bool
}

/// What a chip is for. It holds an id of the release, never anything typed.
enum ChipKind: Hashable, Sendable {
    case tenure
    case budget
    case place(String)
    case feature(FeatureId)
    case tag(TagId)
    case area(String)
    /// Which journey counts, once there are two places to choose between.
    case journeys
    /// The settings nobody chose, as one chip.
    case usual

    /// The part of the search the chip is for, as an edit names it.
    var key: ChipKey? {
        switch self {
        case .tenure: return .tenure
        case .budget: return .budget
        case .place(let id): return .place(id)
        case .feature(let id): return .feature(id)
        case .tag(let id): return .tag(id)
        case .area(let id): return .area(id)
        case .journeys: return .journeys
        case .usual: return nil
        }
    }

    /// True when the chip opens the control the settings hold for it.
    var opensAControl: Bool {
        switch self {
        case .tenure, .budget, .place, .journeys, .feature, .tag: return true
        case .area, .usual: return false
        }
    }
}

struct SearchChip: Hashable, Sendable, Identifiable {
    let kind: ChipKind
    let label: String
    let parts: [ChipPart]
    /// True when the chip as a whole was assumed, or any part of it was.
    let assumed: Bool
    /// The edit that takes it out, where it can be taken out.
    let removal: Operations?

    var id: ChipKind { kind }

    /// True when the word goes after the label, because no one part carries it.
    var assumedAsAWhole: Bool {
        assumed && !parts.contains { $0.assumed }
    }

    /// The whole chip as plain words, as it is drawn and as it is read out.
    var reads: String {
        var said = label
        if assumedAsAWhole { said += " \(SearchCopy.Chips.assumed)" }
        for part in parts {
            said += ", \(part.text)"
            if part.assumed { said += " \(SearchCopy.Chips.assumed)" }
        }
        return said
    }
}

enum SearchChips {
    /// True of a vibe whose recipe holds a figure of recorded crime. Which one does is the API's to say.
    static func holdsRecordedCrime(_ tag: Tag, _ features: [Metric]) -> Bool {
        tag.terms.contains { term in
            features.first { $0.featureId == term.featureId }?.dimension == .crime
        }
    }

    /// What the recipe of a vibe counts that is recorded crime, as one line.
    /// `nil` where it counts none. The name of each part is the API's.
    static func countsCrime(_ tag: Tag, in meta: MetaData) -> String? {
        let parts = tag.terms.compactMap { term in
            meta.features.first { $0.featureId == term.featureId && $0.dimension == .crime }?.label
        }
        guard !parts.isEmpty else { return nil }
        return "\(SearchCopy.crimeCounts): \(parts.joined(separator: "; "))."
    }

    /// Whether a weight counts for anything. A spec keeps an entry of 0 for a
    /// thing a person took off, so that a change of tenure does not bring its
    /// default back (contract section 5.1). Such an entry counts for nothing,
    /// and is drawn as off.
    static func counts(_ weight: Double?) -> Bool {
        (weight ?? 0) > 0
    }

    /// The name of each place of the spec: the release's own name for it, from the
    /// answer that brought the spec. A place no answer named is said to have no name,
    /// and is never shown by a number or by what was typed.
    static func names(of spec: PreferenceSpec, held: [String: String]) -> [String: String] {
        var names: [String: String] = [:]
        for commute in spec.commutes {
            names[commute.placeId] = held[commute.placeId] ?? SearchCopy.Place.noName
        }
        return names
    }

    /// A budget as the person gave it: pounds, and "a month" for a rent.
    static func budget(_ amount: Int, _ tenure: Tenure) -> String {
        let pounds = FormNumbers.grouped(amount)
        return tenure == .rent ? SettingsCopy.Budget.aMonth(pounds) : SettingsCopy.Budget.pounds(pounds)
    }

    /// The chips as the screen draws them: the search's own, and the one that
    /// says which journey counts.
    static func of(_ state: SearchState) -> [SearchChip] {
        withWhichJourneyCounts(
            of(
                state.spec, features: state.meta.features, tags: state.meta.tags, areas: state.areas,
                placeNames: state.placeNames, assumed: state.assumed, guides: state.meta.roughGuides),
            state.spec)
    }

    /// The chips of the search, with the one that says which journey counts
    /// after the places, once there are two of them. With one place there is
    /// nothing to choose between. With two, only one journey feeds the fit
    /// unless the person says otherwise, and until they do the chip is marked
    /// "assumed". It is for the eye alone: a stored search says nothing of it.
    static func withWhichJourneyCounts(_ chips: [SearchChip], _ spec: PreferenceSpec) -> [SearchChip] {
        // A choice the app has no word for is left out.
        guard spec.commutes.count >= 2, let label = CodeCopy.combine(spec.commuteCombine) else {
            return chips
        }
        let from = spec.commuteCombineFrom
        let which = SearchChip(
            kind: .journeys, label: label, parts: [],
            assumed: from == .default || from == .inferred, removal: nil)
        let after = chips.lastIndex { chip in
            if case .place = chip.kind { return true }
            return false
        }
        var drawn = chips
        drawn.insert(which, at: after.map { $0 + 1 } ?? chips.count)
        return drawn
    }

    /// What a vibe that is a rough guide says of itself where it stands in a
    /// search, in one line under the chips: its name, its label and the
    /// sentence that says why. The name, the label and the sentence are the
    /// API's. One line for each such vibe that counts, in the order of the spec.
    static func rough(in spec: PreferenceSpec, meta: MetaData) -> [String] {
        spec.tags.compactMap { weight in
            guard counts(weight.weight),
                let tag = meta.tags.first(where: { $0.tagId == weight.tagId }),
                let told = Vibes.rough(tag, in: meta.roughGuides)
            else { return nil }
            return "\(tag.label): \(Vibes.said(told))"
        }
    }

    static func of(
        _ spec: PreferenceSpec,
        features: [Metric],
        tags: [Tag],
        areas: [AreaSummary],
        placeNames: [String: String],
        assumed: Assumed,
        guides: [RoughGuide] = []
    ) -> [SearchChip] {
        func has(_ key: ChipKey, _ code: AssumptionCode) -> Bool {
            assumed[key]?.contains(code) ?? false
        }
        func notChosen(_ provenance: Provenance) -> Bool {
            provenance == .default || provenance == .inferred
        }
        func chip(
            _ kind: ChipKind, _ label: String, parts: [ChipPart?] = [], assumed: Bool = false,
            removal: Operations? = nil
        ) -> SearchChip {
            // A code the app has no word for makes a part with nothing in it, which is left out.
            let said = parts.compactMap { $0 }.filter { !$0.text.isEmpty }
            return SearchChip(
                kind: kind, label: label, parts: said,
                assumed: assumed || said.contains { $0.assumed }, removal: removal)
        }
        func firmness(_ strictness: Strictness, _ key: ChipKey) -> ChipPart? {
            switch strictness {
            case .hard: return ChipPart(text: SearchCopy.Chips.firm, assumed: has(key, .strictness))
            case .soft: return ChipPart(text: SearchCopy.Chips.flexible, assumed: has(key, .strictness))
            case .unlisted: return nil
            }
        }

        var chips: [SearchChip] = []

        if let tenure = CodeCopy.tenure(spec.tenure) {
            chips.append(
                chip(.tenure, tenure, assumed: notChosen(spec.tenureFrom) || has(.tenure, .tenure)))
        }

        if let amount = spec.budget.amount {
            chips.append(
                chip(
                    .budget, budget(amount, spec.tenure),
                    parts: [
                        CodeCopy.segment(spec.budget.segment).map {
                            ChipPart(text: $0, assumed: has(.budget, .segment))
                        },
                        firmness(spec.budget.strictness, .budget),
                    ],
                    removal: Edits.budgetClear()))
        }

        let names = names(of: spec, held: placeNames)
        for commute in spec.commutes {
            let key = ChipKey.place(commute.placeId)
            chips.append(
                chip(
                    .place(commute.placeId), names[commute.placeId] ?? "",
                    parts: [
                        CodeCopy.mode(commute.mode).map { ChipPart(text: $0, assumed: has(key, .mode)) },
                        ChipPart(
                            text: SearchCopy.Chips.within(commute.maxMinutes),
                            assumed: has(key, .maxMinutes)),
                        firmness(commute.strictness, key),
                    ],
                    removal: Edits.placeRemove(commute.placeId)))
        }

        // An entry of 0 is a thing a person took off. It is said to count for nothing, so
        // that a wish to ignore the high street is never shown as a wish for one, and it
        // has nothing to remove.
        let takenOff = ChipPart(text: SearchCopy.Chips.off, assumed: false)

        for weight in spec.weights where weight.provenance != .default {
            let metric = features.first { $0.featureId == weight.featureId }
            let key = ChipKey.feature(weight.featureId)
            var direction: ChipPart?
            if metric?.polarity == .either {
                switch weight.direction {
                case .more: direction = ChipPart(text: SearchCopy.Chips.more, assumed: has(key, .direction))
                case .less: direction = ChipPart(text: SearchCopy.Chips.fewer, assumed: has(key, .direction))
                case .unlisted: direction = nil
                }
            }
            let on = counts(weight.weight)
            chips.append(
                chip(
                    .feature(weight.featureId), metric?.label ?? weight.featureId.rawValue,
                    parts: on ? [direction] : [takenOff],
                    assumed: weight.provenance == .inferred || has(key, .weight),
                    removal: on ? Edits.featureOff(weight.featureId) : nil))
        }

        for weight in spec.tags {
            let key = ChipKey.tag(weight.tagId)
            let on = counts(weight.weight)
            let tag = tags.first { $0.tagId == weight.tagId }
            // A scale says which of its ends is asked for. Both names are the API's.
            var label = tag?.label ?? weight.tagId.rawValue
            if on, let tag, tag.shape == .scale, let low = tag.lowEnd, let high = tag.highEnd {
                label = SearchCopy.Chips.towards(tag.label, weight.toward == .low ? low : high)
            }
            // A vibe whose recipe holds recorded crime says so wherever it is in a search.
            let crime =
                tag.map { holdsRecordedCrime($0, features) } == true
                ? ChipPart(text: SearchCopy.crimeChip, assumed: false) : nil
            // A vibe that is a rough guide says so on its chip, in the API's own word.
            let rough = tag.flatMap { Vibes.rough($0, in: guides) }
                .map { ChipPart(text: $0.label, assumed: false) }
            chips.append(
                chip(
                    .tag(weight.tagId), label,
                    parts: on ? [rough, crime] : [takenOff],
                    assumed: weight.provenance == .inferred || has(key, .weight),
                    removal: on ? Edits.tagOff(weight.tagId) : nil))
        }

        for rule in spec.areas {
            let how: ChipPart?
            switch rule.rule {
            case .exclude: how = ChipPart(text: SearchCopy.Chips.hidden, assumed: false)
            case .only: how = ChipPart(text: SearchCopy.Chips.only, assumed: false)
            case .unlisted: how = nil
            }
            chips.append(
                chip(
                    .area(rule.areaId),
                    areas.first { $0.areaId == rule.areaId }?.name ?? rule.areaId,
                    parts: [how], removal: Edits.areaClear(rule.areaId)))
        }

        let usual = spec.weights.filter { $0.provenance == .default }.count
        if usual > 0 {
            chips.append(chip(.usual, SearchCopy.Chips.usual(usual), assumed: true))
        }

        return chips
    }
}
